apps/website/prisma/schema.prisma

```diff
@@ -241,4 +249,88 @@ model OutgoingWebhook {
 
   user          User?          @relation(fields: [userId], references: [id])
   giveawayEntry GiveawayEntry?
+
+  @@index([userId], name: "OutgoingWebhook_userId")
+}
+
+model LinkAttachment {
+  id              String @id @default(cuid())
+  type            String
+  name            String
+  title           String
+  alternativeText String
+  caption         String
+  description     String
+  url             String
+
+  showAndTellEntryAttachment ShowAndTellEntryLinkAttachment?
+}
+
+model ShowAndTellEntry {
+  id             String    @id @default(cuid())
+  userId         String?
+  title          String
+  text           String    @db.Text
+  displayName    String?
+  createdAt      DateTime  @default(now())
+  updatedAt      DateTime  @default(now())
+  approvedAt     DateTime?
+  deletedAt      DateTime?
+  seenOnStreamAt DateTime?
+
+  attachments ShowAndTellEntryLinkAttachment[]
+
+  user User? @relation(fields: [userId], references: [id])
+
+  @@index([userId], name: "ShowAndTellEntry_userId")
+}
+
+model ShowAndTellEntryLinkAttachment {
+  id                String           @id @default(cuid())
+  entryId           String
+  attachmentType    String
+  linkAttachmentId  String?          @unique
+  imageAttachmentId String?          @unique
+  entry             ShowAndTellEntry @relation(fields: [entryId], references: [id])
+  linkAttachment    LinkAttachment?  @relation(fields: [linkAttachmentId], references: [id])
+  imageAttachment   ImageAttachment? @relation(fields: [imageAttachmentId], references: [id])
+
+  @@index([entryId], name: "ShowAndTellEntryLinkAttachment_entryId")
+}
+
+model FileStorageObject {
+  id         String    @id @default(cuid())
+  key        String    @unique
+  name       String
+  type       String
+  prefix     String
+  acl        String
+  createdAt  DateTime  @default(now())
+  uploadedAt DateTime?
+  deletedAt  DateTime?
+  expiresAt  DateTime?
+
+  imageAttachment ImageAttachment?
+  imageMetadata   ImageMetadata?
+}
+
+model ImageMetadata {
+  mimeType            String
+  width               Int
+  height              Int
+  fileStorageObjectId String            @unique
+  fileStorageObject   FileStorageObject @relation(fields: [fileStorageObjectId], references: [id])
+}
+
+model ImageAttachment {
+  id                             String                          @id @default(cuid())
+  name                           String
+  title                          String
+  alternativeText                String
+  caption                        String
+  description                    String
+  url                            String
+  fileStorageObjectId            String                          @unique
+  fileStorageObject              FileStorageObject               @relation(fields: [fileStorageObjectId], references: [id])
+  showAndTellEntryLinkAttachment ShowAndTellEntryLinkAttachment?
```

Comment:
Through this relation, isn't there a relation to a LinkAttachment which already contains much of the metadata (title, description, etc.) that this model seems to be duplicating? Or am I misunderstanding the schema?

Reply:
It's complicated and probably over-engineered, but there should not be any duplications.

- if a file upload is started, a `FileStorageObject` row also gets created to keep a reference for cleanup etc.
- an `ShowAndTellEntry` can have many attachments that are either image or links
- to implement this polymorphic m:m relation I've added the `ShowAndTellEntryLinkAttachment` table that should have been named `ShowAndTellEntryAttachment`. Prisma makes it a bit harder to keep the m:m table simple.
- link attachments and image attachments then have all the additional fields a files may have. to keep those entities generic I over-engineered them and added a bunch of fields we currently do not use. I basically copied what the old page did. but they did not really get used there either
- if a new image attachment is created, it is connected with the uploaded file storage object if it is valid
- at this point we also check the image to be valid and extract some info (mime type, width, height)
- if the image is not valid it gets deleted
- if the image meta data is valid, it gets written to the `ImageMetadata` connected to the `FileStorageObject`
- the `FileStorageObject` is generic though and can be any file type so I kept the metadata in a separate table

![alveus_attachments_er](https://user-images.githubusercontent.com/684458/222857994-80503d1b-a083-487c-8ce1-0afc2dea7862.png)

Reply:
I renamed the m:m relation table from `ShowAndTellEntryLinkAttachment` to `ShowAndTellEntryAttachment`

Reply:
(fwiw, it might be worth dropping this explanation into a docs MD file somewhere in the repo, for future travellers?)

That diagram makes it much easier to follow -- I'd got tripped up by the fact the Prisma schema defines the relationship on both models, not just the "parent" model (e.g. the EntryAttachment), so was thinking an ImageAttachment was tied to a LinkAttachment somehow.